orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1781,22 +1752,18 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
   }
   attrs.push_back(MakeAttribute("output_tensor_ranks", output_tensor_ranks));
 
-  std::vector<int64_t> output_tensor_requires_grads;
-  for (const auto requires_grad : src_attrs["input_tensor_requires_grads"].ints()) {
-    output_tensor_requires_grads.push_back(requires_grad);
-  }
-  attrs.push_back(MakeAttribute("output_tensor_requires_grads", output_tensor_requires_grads));
-
   std::vector<ArgDef> input_args;
   // Put Python context generated by PythonOp.
   input_args.push_back(O(0));
   // Put other outputs.
   for (int i = 1; i < GetSrcNodeOutputSize(); ++i) {
-    if (input_tensor_requires_grads.at(i)) {
+    if (src_attrs["output_tensor_requires_grads"].ints().Get(i - 1)) {
       // Only add FW outputs which
       //  1. are tensors,
       //  2. needs gradients (requires_grad=True in Pytorch).
       input_args.push_back(GO(i));
+    } else {
+      input_args.push_back(ArgDef());
     }
   }
 
```

Comment:
Stashing all the outputs for backward computation could contribute to higher peak memory... 
I wonder if the seg fault is related to the ctx.save_for_backward(output_tensor)? 

This should be marked as an to-investigate issue.

Reply:
The seg fault happens when ctx is teared down before PythonOpGrad runs. To make sure ctx is not destroyed, at least one of those forward outputs remains before PythonOpGrad kernels completes. So you are right, we should not keep all outputs if we could identify which part of them contributes to backward computations. I will follow up on it.

Reply:
Added a TODO